content/articles/automation_lessons.md

```diff
@@ -0,0 +1,110 @@
+title: Automation Lessons
+tags: software engineering, collaboration
+author: cdent
+created: 2025-02-26
+
+```intro
+I initially wrote the following for work colleagues, but sharing more widely
+in case it might apply elsewhere.
+
+A lesson from long ago that has stuck with me and been proven valuable over and
+over:
+
+> When creating software, ensure there are working harnesses for testing,
+> installation, and upgrade first.
+```
+
+Over the years as development technologies have matured the lesson has evolved
+to assert the harnesses must be _automated_ and that the automation _must run
+before changes are merged_.
+
+We've all probably learned this main lesson and the various side lessons many
+times over, but frequent review is required because...
+
+...Throughout my experience with our previous teams we've frequently ignored
+this lesson for a variety of reasons, all of which are at least inconvenient
+but most of the time invalid:
+
+1. We're in too much of a hurry to add tests, we need to get this out now.
+2. This is too complicated to automate.
+3. This is too time consuming to run in pull request workflows.
+4. This is too resource consuming to run in pull request workflows.
```

Comment:
Note to self: cdent doesn't do hyphens. Just don't mention it.

Reply:
I'm not sure what you're saying, but my answer is readability; forward-looking hints, essentially.